### ISSUE TITLE
Implement balloon perspective descriptions

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/balloons-and-static-electricity#154]
Seems like there really wan't an issue for this yet.  This issue will just track progress of implementing dynamic descriptions on the balloon which are dependent on balloon position, charge, whether or not it has focus, whether or not it is moving, etc...

For description design, see https://docs.google.com/spreadsheets/d/1BiXFN2dRWfsjqV2WvKAXnZFhsk0jCxbnT0gkZr_T5T0/edit?ts=568067c0#gid=931000658

EDIT - Adding checklist to track progress of the major milestones for this issue.
- [ ]  Single Balloon Perspective 1
- [x] Single Balloon Perspective 2
- [x] Single Balloon Perspective 3
- [ ] Interaction alerts

**Comment by terracoda** [comment]
Thanks @jessegreenberg for opening this issue to track updates to descriptions.

**Comment by terracoda** [comment]
@jessegreenberg, I did write up summary descriptions for when a user presses the resets. These are in Tab 6. I have 2 columns. The second column is the super brief version.

**Comment by jessegreenberg** [comment]
Thanks @terracoda, Ill check it out!

**Comment by jessegreenberg** [comment]
I added comments in the google doc, but I have two questions:

- What should the descriptions be when the balloon is released from the right most side of the play area when the wall is not there.

- When is the 'Resting State/Location at Wall' description announced for a neutral balloon?

**Comment by jessegreenberg** [comment]
Above commit implements the balloon location descriptions and alerts when balloon is released.

**Comment by jessegreenberg** [comment]
Above commits should complete perspective 3 (description upon release).

**Comment by terracoda** [comment]
@jessegreenberg, this is great. I will look at the description on Friday.

**Comment by terracoda** [comment]
Hi @jessegreenberg,
Thanks for all your hard work. 
Is the latest instance still:
http://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.2.0-reader.5/balloons-and-static-electricity_en.html?reader

**Comment by terracoda** [comment]
@jessegreenberg, I have a completed first draft of items in a Dynamic Scene Summary. They are in the google doc that contains all descriptions. They are on the PDOM tab. I created several rows for the Scene Summary to break it up, but really it is all part of the same thing.
[BASE Labels and Descriptions](https://docs.google.com/spreadsheets/d/1BiXFN2dRWfsjqV2WvKAXnZFhsk0jCxbnT0gkZr_T5T0/edit#gid=0)

I now see the Scene Summary as having the following static and dynamic pieces:

1. Static: H2
2. Static: General opening paragraph
3. Dynamic: State of the room expressed as a list of 3 bullets (maybe 4) providing a big picture view of the room, or play area. These 3 items are:
  1. Number of objects in the room. 
  2. Balloon position, or positions of balloons.
  3. Description of amount of charge on the different objects. 
4. Preferably Static: Navigation cue. 
Note that we may need a couple of versions of the navigation cue (or less specific one) because "Tab to play with balloon" is no longer a true direction once virtual cursor focus has moved to another focusable item in the sim.

Hopefully we can discuss this on Tuesday's meeting, or if need be set up another meeting this week. July 4th must be a holiday for you. Have a good one.

**Comment by jessegreenberg** [comment]
> https://github.com/phetsims/balloons-and-static-electricity/issues/154#issuecomment-230111595

@terracoda yes, that is still the latest deployed version.  I wanted to get a bit farther with dynamic descriptions before deploying another.  But I will deploy the next dev version soon.

> https://github.com/phetsims/balloons-and-static-electricity/issues/154#issuecomment-230169737

Thanks @terracoda, this looks great!  I just saw the updates to the design document.  I really like the idea of breaking up the description into a few static/dynamic pieces.

**Comment by terracoda** [comment]
@jessegreenberg Perspective one is a little less clear since we changed the interaction of grabbing the balloon. I will need to review this a bit more. I think Perspective 2 and 3 are more clear. 

- Perspective 2: the balloon has been grabbed and the user is figuring out how to move it or is actively moving it.
- Perspective 3: is the description upon release.

Perspective 1 was originally for the description upon focus, but not yet grabbed. This state does not really exist any more, exactly like that. The Grab Balloon button is involved. I need to look at this a bit more. Can you leave until I see the latest release?

**Comment by jessegreenberg** [comment]
> Perspective 1 was originally for the description upon focus, but not yet grabbed. This state does not really exist any more, exactly like that. The Grab Balloon button is involved. I need to look at this a bit more. Can you leave until I see the latest release?

Sure @terracoda, I will hold off.  While reading perspective 1, was that it was for the balloon when it is not grabbed and not in focus, but found by the virtual cursor.  So it is the description of the balloon in its 'stationary' state.  Would that be correct?

**Comment by terracoda** [comment]
@jesse, Yes, __Perspective 1__ is a description of the balloon in its _'stationary' ungrabbed state_ - that is exactly right. This is the description of the balloon in the Parallel DOM if the user uses the cursor keys to get to it. When we changed the interaction (added the Grab Yellow/ Green Balloon Button), the result was that we hid the balloon from the user.

What I think we need to do is hide the interactive description of the balloon from the user, but find a way to provide its actual description (as described in Perspective 1) which would be accessible with the cursor keys. It is essential that the 'stationary balloon' description NOT include any description of how to move the balloon as moving the balloon is only possible when it is active (i.e. grabbed).

If we can make the _stationary description_ available via the cursor keys, what we might need an _initial hint_ to look for the grab button as part of the stationary description. It would be great to have _initial hints_ disappear once a user executes the interaction (i.e. executing the interaction hopefully means they learned it).

I reviewed Perspective 1 and it all looks good. The only thing missing would be text for an initial hint to find the grab button.

**Comment by terracoda** [comment]
@jessegreenberg, so the bottom part of the __Perspective 1__ google sheet describes the balloon's description as it would be when the user presses the grab button to grab the balloon. This interaction dynamically adds the descriptions of the balloon's nearby object and dynamically changes the description for the navigation cue (or initial interaction or navigation hints).

**Comment by terracoda** [comment]
@jessegreenberg, just had an idea about how to possibly make navigation cues in to initial interaction hints. This idea might be most useful for the grab balloon button.  
Idea: use aria-described-by on the controls on the first go around, but then

**Comment by terracoda** [comment]
@jessegreenberg, fyi, I updated the previous comment as I posted it prematurely - an idea to keep in mind anyways.

**Comment by jessegreenberg** [comment]
@terracoda regarding https://github.com/phetsims/balloons-and-static-electricity/issues/154#issuecomment-233147970, this sounds interesting, we could do this.  Up to you for whether or not you would like try it.

**Comment by jessegreenberg** [comment]
Big commit above begins some initial work on new balloon alerts.  In addition, it generalizes the accessible drag handler for use in other places.

**Comment by jessegreenberg** [comment]
Above commit creates a common type for accessible nodes.  Button to grab balloon and accessible drag hanlders extend this.

**Comment by jessegreenberg** [comment]
@terracoda has provided a list of interaction walkthroughs, outlining description cases.

https://docs.google.com/spreadsheets/d/1b6zfzdbRwYDoHA5ecGEp8atx4moFJOf7aYXNGDFF1bE/edit?usp=sharing

This has been very helpful in exemplifying description cases.

**Comment by terracoda** [comment]
@jessegreenberg thanks. This is the better issue for the mini walkthroughs.

**Comment by terracoda** [comment]
@jessegreenberg, I've added interaction alerts (based on your idea) that can be used to find remaining charges on the sweater. These are located in the Interaction Alerts tab in [BASE Labels & Descriptions](https://docs.google.com/spreadsheets/d/1BiXFN2dRWfsjqV2WvKAXnZFhsk0jCxbnT0gkZr_T5T0/edit#gid=232870371). See line 20.
For an initial example of these in action see scenario number 4 the [mini cognitive walkthroughs sheet](https://docs.google.com/spreadsheets/d/1b6zfzdbRwYDoHA5ecGEp8atx4moFJOf7aYXNGDFF1bE/edit#gid=0)

**Comment by terracoda** [comment]
@jessegreenberg, actually the interaction alerts for "Rubbing with no charges" is at row 19 on INTERACTION ALERTS Tab in [BASE Labels & Descriptions](https://docs.google.com/spreadsheets/d/1BiXFN2dRWfsjqV2WvKAXnZFhsk0jCxbnT0gkZr_T5T0/edit#gid=232870371). Just look for the heading "Rubbing with no charges." Updated previous comment.

**Comment by terracoda** [comment]
@jessegreenberg I review the PDOM hierarchy Tab as well in [BASE Labels & Descriptions](https://docs.google.com/spreadsheets/d/1BiXFN2dRWfsjqV2WvKAXnZFhsk0jCxbnT0gkZr_T5T0/edit#gid=232870371). I think it is much clearer now.

I marked changed cells in with a light red (pinkish) background color.

**Comment by terracoda** [comment]
@jessegreenberg, I added two general Interaction Alerts for the Balloon Reset button. Not sure we need more at this point. 
- Reset Balloon (single): Sweater and balloon reset to net neutral charge. Balloon, in center of play area.
- Reset Balloons (two): Sweater and balloons reset to net neutral charge. Balloons, in center of play area.
Let's see if a simple approach works :-)

**Comment by jessegreenberg** [comment]
Excellent, thanks @terracoda I am looking forward to working on the changes and new alerts for discoverability! :)

**Comment by jessegreenberg** [comment]
Large commit above implements the descriptions for ~80% of the dragging alerts in the walk througs.  Next steps will be to finish https://docs.google.com/spreadsheets/d/1b6zfzdbRwYDoHA5ecGEp8atx4moFJOf7aYXNGDFF1bE/edit#gid=0
then send the descriptions to the aria-live elements via AriaHerald so that they are announced by AT.

**Comment by terracoda** [comment]
@jessegreenberg, this looks like an amazing amount of work! Thank you!

**Comment by jessegreenberg** [comment]
My pleasure @terracoda, sorry the deploy is a bit later than expected.  Latest version sent in email, which described my experiences running in various platforms.  Specific issues should be enumerated in new github tickets.

I think this issue can now be closed.  We decided to not continue with perspectives tabs, and to base design off of PDOM hierarchy, interaction alerts, play area grid, example interactions and so on.  Review issues and remaining tasks will be discussed in new issues.  @terracoda if you disagree, feel free to reopen this issue.
